File: src/main/java/com/keygenqt/mylibrary/base/JWTAuthorizationFilter.java
/*
 * Copyright 2020 Vitaliy Zarubin
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.keygenqt.mylibrary.base;

import com.keygenqt.mylibrary.config.WebSecurityConfig;
import com.keygenqt.mylibrary.models.User;
import com.keygenqt.mylibrary.models.UserToken;
import com.keygenqt.mylibrary.models.repositories.UserRepository;
import com.keygenqt.mylibrary.models.repositories.UserTokenRepository;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import org.jetbrains.annotations.NotNull;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.filter.OncePerRequestFilter;

import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.util.ArrayList;
import java.util.Date;
import java.util.LinkedHashMap;

public class JWTAuthorizationFilter extends OncePerRequestFilter {

    public static final String ACCEPT_LANGUAGE = "Accept-Language";
    public static final String HEADER = "Authorization";

    private static final String PREFIX = "Bearer ";

    private final UserRepository repositoryUser;
    private final UserTokenRepository repositoryUserToken;

    public JWTAuthorizationFilter(UserRepository repositoryUser, UserTokenRepository repositoryUserToken) {
        this.repositoryUser = repositoryUser;
        this.repositoryUserToken = repositoryUserToken;
    }

    public static String getJWTToken(String login, String role) {
        return JWTAuthorizationFilter.PREFIX + Jwts
                .builder()
                .setId("JWT")
                .setSubject(login)
                .claim("authorities", new ArrayList<SimpleGrantedAuthority>() {{
                    add(new SimpleGrantedAuthority(role));
                }})
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + 2592000000L /* month */))
                .signWith(SignatureAlgorithm.HS512, WebSecurityConfig.SECRET_KEY.getBytes())
                .compact();
    }

    @Override
    public void doFilterInternal(@NotNull HttpServletRequest request, @NotNull HttpServletResponse response, @NotNull FilterChain chain) {

        String authorization = request.getHeader(HEADER);
        String language = request.getHeader(ACCEPT_LANGUAGE);

        try {
            if (authorization != null && language != null && authorization.startsWith(PREFIX)) {
                Claims claims = validateToken(authorization, language);
                if (claims != null && claims.get("authorities") != null) {
                    setUpSpringAuthentication(claims);
                    chain.doFilter(request, response);
                    return;
                }
            }
            SecurityContextHolder.clearContext();
            chain.doFilter(request, response);
        } catch (Exception ex) {
            System.out.println(ex.getMessage());
        }
    }

    private Claims validateToken(String authorization, String language) {
        UserToken token = repositoryUserToken.findByToken(authorization);
        if (token == null) {
            return null;
        }

        User user = repositoryUser.findByIdActive(token.getUserId());
        if (user == null || !user.getEnabled()) {
            return null;
        }

        update(token, language);

        return Jwts.parser()
                .setSigningKey(WebSecurityConfig.SECRET_KEY.getBytes())
                .parseClaimsJws(authorization.replace(PREFIX, ""))
                .getBody();
    }

    private void setUpSpringAuthentication(Claims claims) {
        var authorities = claims.get("authorities");
        if (authorities instanceof ArrayList && !((ArrayList<?>) authorities).isEmpty()) {
            var item = ((ArrayList<?>) authorities).get(0);
            if (item instanceof LinkedHashMap && ((LinkedHashMap<?, ?>) item).containsKey("authority")) {
                var authority = ((LinkedHashMap<?, ?>) item).get("authority");
                SecurityContextHolder.getContext().setAuthentication(
                        new UsernamePasswordAuthenticationToken(claims.getSubject(), null, new ArrayList<SimpleGrantedAuthority>() {{
                            add(new SimpleGrantedAuthority((String) authority));
                        }})
                );
            }
        }
    }

    private void update(UserToken userToken, String language) {
        var isUpdate = false;
        if (!userToken.getLanguage().equals(language)) {
            userToken.setLanguage(language);
            isUpdate = true;
        }
        if (userToken.getUpdatedAt().getTime() < (new Date().getTime() - 86400000)) {
            userToken.setUpdatedAt(new Date());
            isUpdate = true;
        }
        if (isUpdate) {
            repositoryUserToken.save(userToken);
        }
    }
}

File: src/main/kotlin/com/keygenqt/mylibrary/config/JWTAuthorizationFilter.kt
/*
 * Copyright 2020 Vitaliy Zarubin
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.keygenqt.mylibrary.config

import com.keygenqt.mylibrary.models.*
import com.keygenqt.mylibrary.models.repositories.*
import io.jsonwebtoken.*
import org.springframework.http.HttpStatus.*
import org.springframework.security.authentication.*
import org.springframework.security.core.authority.*
import org.springframework.security.core.context.*
import org.springframework.web.filter.*
import org.springframework.web.server.*
import java.util.*
import javax.servlet.*
import javax.servlet.http.*

class JWTAuthorizationFilter(
    private val repositoryUser: UserRepository,
    private val repositoryUserToken: UserTokenRepository
) : OncePerRequestFilter() {

    companion object {
        const val ACCEPT_LANGUAGE = "Accept-Language"
        const val HEADER = "Authorization"
        const val PREFIX = "Bearer "

        fun getJWTToken(login: String, role: String): String {
            return "Bearer ${
                Jwts
                    .builder()
                    .setId("JWT")
                    .setSubject(login)
                    .claim("authorities", listOf(SimpleGrantedAuthority(role)))
                    .setIssuedAt(Date(System.currentTimeMillis()))
                    .setExpiration(Date(System.currentTimeMillis() + 2592000000 /* month */))
                    .signWith(SignatureAlgorithm.HS512, WebSecurityConfig.SECRET_KEY.toByteArray())
                    .compact()
            }"
        }
    }

    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, chain: FilterChain) {
        try {
            if (checkJWTToken(request)) {
                val claims = validateToken(request)
                if (claims["authorities"] != null) {
                    setUpSpringAuthentication(claims)
                    chain.doFilter(request, response)
                    return
                }
            }
        } catch (ex: Exception) {
        }
        SecurityContextHolder.clearContext()
        chain.doFilter(request, response)
    }

    private fun validateToken(request: HttpServletRequest): Claims {
        val token = request.getHeader(HEADER)
        val language = request.getHeader(ACCEPT_LANGUAGE)
        repositoryUserToken.findByToken(token)?.let { userToken ->
            repositoryUser.findByIdActive(userToken.userId)?.let { user ->
                if (user.enabled) {
                    update(userToken, language)
                    return Jwts.parser()
                        .setSigningKey(WebSecurityConfig.SECRET_KEY.toByteArray())
                        .parseClaimsJws(token.replace(PREFIX, ""))
                        .body
                } else {
                    throw ResponseStatusException(UNAUTHORIZED, "Authorization failed. User disabled.")
                }
            }
        }
        throw ResponseStatusException(UNAUTHORIZED, "Authorization failed")
    }

    private fun setUpSpringAuthentication(claims: Claims) {
        claims["authorities"]?.let {
            if (it is ArrayList<*>) {
                it.first()?.let { item ->
                    if (item is LinkedHashMap<*, *>) {
                        item["authority"]?.let { authority ->
                            SecurityContextHolder.getContext().authentication =
                                UsernamePasswordAuthenticationToken(claims.subject, null, listOf(SimpleGrantedAuthority(authority.toString())))
                        }
                    }
                }
            }
        }
    }

    private fun checkJWTToken(request: HttpServletRequest): Boolean {
        request.getHeader(HEADER)?.let {
            return it.startsWith(PREFIX)
        }
        return false
    }

    private fun update(userToken: UserToken, language: String) {
        var isUpdate = false
        if (language != userToken.language) {
            userToken.language = language
            isUpdate = true
        }
        if (userToken.updatedAt.time < (Date().time - 86400000)) {
            userToken.updatedAt = Date()
            isUpdate = true
        }
        if (isUpdate) {
            repositoryUserToken.save(userToken)
        }
    }
}